### PR TITLE
Add hasStandaloneWarmth for OnyxWarmthController devices

### PR DESCRIPTION
**Pull request by mergen3107**
@pazos 
These are the changes I did (https://github.com/koreader/koreader/pull/8120#issuecomment-1636970460)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/android-luajit-launcher/426)

### CODE REVIEW

**Comment by mergen3107** [comment]
Do not merge yet please, needs testing.

**Comment by pazos** [comment]
> (P.S. I am trying now to put these file changes to a branch so you can see better)

Thanks. Way easier for me :)

> `Abort message: 'JNI DETECTED ERROR IN APPLICATION: JNI CallBooleanMethod called with pending exception java.lang.NoSuchMethodError: no non-static method "Lorg/koreader/launcher/MainActivity;.hasStandaloneWarmth()Z"`

> What am I doing wrong?

Nothing wrong, my comment wasn't exhaustive. You'll need to add the method to https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/LuaInterface.kt and override it in https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/MainActivity.kt

It (mainactivity) will look like:

```kotlin
override fun hasStandaloneWarmth(): Boolean {
    return device.lights.hasStandaloneWarmth()
}
```

**Comment by mergen3107** [comment]
Thank you @pazos !
KOReader was able to open with these changes.
I am able to turn off the frontlight (cold and warm) via a gesture.

However, when I turn it back on, only cold light turns on. I am trying to investigate this.

**Comment by mergen3107** [comment]
> However, when I turn it back on, only cold light turns on. I am trying to investigate this.

https://github.com/koreader/koreader/pull/11275

`android-luajit` side of things (this PR) is correct.
The mentioned PR is in discussion.

@pazos 
Please review this PR. It should be generalized enough so that we don't need to add devices one by one to make these changes apply - as long as they properly refer to this OnyxWarmthController.

Thank you for all your help!

File: app/src/main/java/org/koreader/launcher/device/lights/OnyxSdkLightsController.kt [hunk 1]
```diff
@@ -76,6 +76,9 @@ class OnyxSdkLightsController : LightsInterface {
     override fun enableFrontlightSwitch(activity: Activity): Int {
         return 1
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/OnyxSdkLightsController.kt:79]: newline before the new method

File: app/src/main/java/org/koreader/launcher/device/lights/OnyxColorController.kt [hunk 2]
```diff
@@ -76,4 +76,7 @@ class OnyxColorController : LightsInterface {
     override fun getMaxBrightness(): Int {
         return BRIGHTNESS_MAX
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/OnyxColorController.kt:79]: newline before the new method

File: app/src/main/java/org/koreader/launcher/device/lights/OnyxC67Controller.kt [hunk 3]
```diff
@@ -69,4 +69,7 @@ class OnyxC67Controller : LightsInterface {
     override fun getMaxBrightness(): Int {
         return BRIGHTNESS_MAX
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/OnyxC67Controller.kt:72]: newline before the new method

File: app/src/main/java/org/koreader/launcher/device/lights/GenericController.kt [hunk 4]
```diff
@@ -88,4 +88,7 @@ class GenericController : LightsInterface {
     override fun getMaxBrightness(): Int {
         return BRIGHTNESS_MAX
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/GenericController.kt:91]: newline before the new method

File: app/src/main/java/org/koreader/launcher/device/lights/BoyueS62RootController.kt [hunk 5]
```diff
@@ -84,4 +84,7 @@ class BoyueS62RootController : LightsInterface {
     override fun getMaxBrightness(): Int {
         return BRIGHTNESS_MAX
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/BoyueS62RootController.kt:87]: newline before the new method

File: app/src/main/java/org/koreader/launcher/LuaInterface.kt [hunk 6]
```diff
@@ -48,6 +48,7 @@ interface LuaInterface {
     fun hasLights(): Boolean
     fun hasNativeRotation(): Boolean
     fun hasOTAUpdates(): Boolean
+    fun hasStandaloneWarmth(): Boolean
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/LuaInterface.kt:51]: sorted, please

File: app/src/main/java/org/koreader/launcher/device/lights/TolinoNtxController.kt [hunk 7]
```diff
@@ -107,4 +107,7 @@ class TolinoNtxController : Ioctl(), LightsInterface {
     override fun getMaxBrightness(): Int {
         return BRIGHTNESS_MAX
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/TolinoNtxController.kt:110]: newline before the new method

File: app/src/main/java/org/koreader/launcher/device/lights/TolinoRootController.kt [hunk 8]
```diff
@@ -151,4 +151,7 @@ class TolinoRootController : LightsInterface {
     override fun getMaxBrightness(): Int {
         return BRIGHTNESS_MAX
     }
+    override fun hasStandaloneWarmth(): Boolean {
```
Comment (pazos) [inline app/src/main/java/org/koreader/launcher/device/lights/TolinoRootController.kt:154]: newline before the new method

**Comment by pazos** [comment]
> Please review this PR. It should be generalized enough so that we don't need to add devices one by one to make these changes apply.

That's the entire point of the device abstraction ;)

PR is fine by me, please address specific comments.

**Comment by mergen3107** [comment]
Thank you!
While adding my newlines, I noticed some methods didn't have them, so I added those newlines too, for consistency.

**Review by pazos — APPROVED** [review]

**Pull request merged**